File: Algorithms/Swift/Medium/2_Add_Two_Numbers.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */
class Solution {
    func addTwoNumbers(l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        if l1 == nil {
            return l2
        }
        if l2 == nil {
            return l1
        }

        let dummy = ListNode(0)
        var head: ListNode? = dummy, carry = 0, list1 = l1, list2 = l2
        while carry != 0 || list1 != nil || list2 != nil {
            var sum = 0

            if list1 != nil {
                sum += list1!.val
                list1 = list1!.next
            }
            if list2 != nil {
                sum += list2!.val
                list2 = list2!.next
            }
            sum += carry

            carry = sum / 10
            sum %= 10

            let node = ListNode(sum)
            head!.next = node
            head = head!.next
        }

        return dummy.next
    }
}

File: Algorithms/JavaScript/Medium/2_Add_Two_Numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    var dummyNode = new ListNode(0)

    var node = dummyNode, carry = 0
    while (l1 !== null || l2 !== null || carry !== 0) {
        var sum = 0

        if (l1 !== null) {
            sum += l1.val
            l1 = l1.next
        }
        if (l2 !== null) {
            sum += l2.val
            l2 = l2.next
        }
        sum += carry

        carry = Math.floor(sum / 10)
        sum %= 10

        var newNode = new ListNode(sum)
        node.next = newNode
        node = node.next
    }

    return dummyNode.next
};
